### PR TITLE
[5.x][RFC] Plugins loading concept. Creating the plugin instance on demand (when the respective event is triggered)

### PR DESCRIPTION
**Pull request by Fedik**
### Summary of Changes

This is concept of loading the plugins on demand.
The idea (not new) is to instantiate the plugin only when the event is triggered, thus reduce App initialisation and execution in general. This can be achieved for plugins that implements `SubscriberInterface`.

The PR contains an examples based on: LanguageFilter, Debug and GuidedTours plugins.

**It works as folowing:**
`PluginHelper::importPlugin()` reading the plugin event with `getSubscribedEvents()`, and creates a clousore for each.
When the event is triggered the closure is looking for the plugin instance, and executes the event method.

Instead of use $app->bootPlugin() it introduce $app->getPluginContainer(), which allows to retrieve the plugin information without booting it in first place.

**New options for plugin service:**
- `plugin.information` Define the plugin class name, and implemented interfaces. This is important to be able to get the class name without booting whole plugin. (Should be probably some kind of configurable class, that can read this information from XML automaticaly, or set by developer. Need to discuss, please share your ideas)
- `plugin.executeValidation` Validate if the plugin should be executed in current runtime. This solves long standing problem, when we have to run `$app->isClinet('potato')` in many events. (Should probably also be some kind of configurable class. Need to discuss, please share your ideas)

The PR uses a simple arrays for new options, however it is only for purpuse of demonstration. Some of these things need to be updated to to proper Class/Interface.

**impact on existing plugins:**

Zero, Null, Nulo, no backward compatibility break. All existing plugins will work as before.

### Testing Instructions

The PR can be merged without testing.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed
- [ ] Pull Request link for manual.joomla.org: TBD
- [ ] No documentation changes for manual.joomla.org needed

Please share your opinions :beers:

### CODE REVIEW

File: libraries/src/Plugin/PluginHelper.php [hunk 1]
```diff
@@ -189,7 +192,8 @@ public static function importPlugin($type, $plugin = null, $autocreate = true, D
             // Get the specified plugin(s).
             for ($i = 0, $t = \count($plugins); $i < $t; $i++) {
                 if ($plugins[$i]->type === $type && ($plugin === null || $plugins[$i]->name === $plugin)) {
-                    static::import($plugins[$i], $autocreate, $dispatcher);
+                    //static::import($plugins[$i], $autocreate, $dispatcher);
```
Comment (brianteeman) [inline libraries/src/Plugin/PluginHelper.php:195]:
```suggestion
```

Comment (Fedik) [reply libraries/src/Plugin/PluginHelper.php:195]: I keeping it as reminder to deprecate `import()` method later on :wink:

Comment (brianteeman) [reply libraries/src/Plugin/PluginHelper.php:195]: so please add a comment to that effect

File: libraries/src/Plugin/PluginHelper.php [hunk 2]
```diff
@@ -205,6 +209,75 @@ public static function importPlugin($type, $plugin = null, $autocreate = true, D
...
+                $dispatcher->addSubscriber($instance);
+            } else {
+                // Add a listener, which will instantiate the plugin when the event will dispatch
+                foreach ($class::getSubscribedEvents() as $eventName => $handler) {
+                    $dispatcher->addListener($eventName, function ($event) use ($container, $handler, &$instance) {
+                        $instance = $instance ?? $container->get(PluginInterface::class);
+                        $caller   = \is_array($handler) ? [$instance, $handler[0]] : [$instance, $handler];
+
+                        return $caller($event);
+                    }, \is_array($handler) ? $params[1] ?? Priority::NORMAL : Priority::NORMAL);
```
Comment (Denitz) [inline libraries/src/Plugin/PluginHelper.php]: @Fedik Maybe it's `$handler[1]` but not `$params[1]`?

Comment (Fedik) [reply libraries/src/Plugin/PluginHelper.php]: yeap, I will fix it, thanks

File: libraries/src/Extension/ExtensionManagerTrait.php [hunk 3]
```diff
@@ -100,23 +110,44 @@ public function bootPlugin($plugin, $type): PluginInterface
...
+     * Retrieve container for a plugin.
+     *
+     * @param   string  $plugin  The plugin name
+     * @param   string  $type    The type of the plugin
+     *
+     * @return  ContainerInterface
+     *
+     * @since  __DEPLOY_VERSION__
+     */
+    public function getPluginContainer($plugin, $type): ContainerInterface
```
Comment (Fedik) [inline libraries/src/Extension/ExtensionManagerTrait.php:122]: **Note:** This should not be public in final implementation. Made like that to "make it work" purpuse only (and for demonstartion).
Probably need a standalone ExtesionManager class.

File: libraries/src/Plugin/PluginHelper.php [hunk 4]
```diff
@@ -205,6 +209,75 @@ public static function importPlugin($type, $plugin = null, $autocreate = true, D
...
+
+            if ($eager) {
+                // Eager plugin instance
+                $instance = $container->get(PluginInterface::class);
+                $dispatcher->addSubscriber($instance);
+            } else {
+                // Add a listener, which will instantiate the plugin when the event will dispatch
+                foreach ($class::getSubscribedEvents() as $eventName => $handler) {
+                    $dispatcher->addListener($eventName, function ($event) use ($container, $handler, &$instance) {
+                        $instance = $instance ?? $container->get(PluginInterface::class);
```
Comment (Fedik) [inline libraries/src/Plugin/PluginHelper.php:263]: **Note:** It probably should still use bootPlugin() to get the plugin instance, to keep API consistent.

**Pull request closed without merging**